src/drivers/efi/capsules.c

```diff
@@ -0,0 +1,759 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#include <acpi/acpi.h>
+#include <boot/coreboot_tables.h>
+#include <bootmem.h>
+#include <bootstate.h>
+#include <cbmem.h>
+#include <console/console.h>
+#include <cpu/x86/pae.h>
+#include <drivers/efi/efivars.h>
+#include <drivers/efi/capsules.h>
+#include <memrange.h>
+#include <string.h>
+#include <stdio.h>
+#include <smmstore.h>
+#include <types.h>
+
+#include <Uefi/UefiSpec.h>
+#include <Guid/GlobalVariable.h>
+#include <Guid/FmpCapsule.h>
+#include <IndustryStandard/WindowsUxCapsule.h>
+
+/*
+ * Overview
+ *
+ * SG stands for scatter-gather.  SG list consists of SG blocks that describe a
+ * potentially discontinuous sequence of memory blocks while not necessarily
+ * lying in continuous memory themselves.
+ *
+ * SG list is basically a linked list of arrays of block descriptors (SG
+ * blocks).  Each of SG blocks can be:
+ *  - a data block, which points to capsule's data
+ *  - a continuation block, which says where other SG blocks are to be found
+ *  - end-of-list block, which indicates there are no more blocks
+ *
+ * Each of the CapsuleUpdateData* EFI variables point to some SG list which
+ * might contain one or more update capsules.  SG blocks never contain data of
+ * more than one of the capsules.  Boundary between capsules in an SG list is
+ * determined by parsing capsule headers and counting amount of data seen so
+ * far.
+ *
+ * There can be multiple CapsuleUpdateData* variables (CapsuleUpdateData,
+ * CapsuleUpdateData1, etc.) in which case their SG lists are chained together
+ * after sanity checks.
+ */
+
+/* This should be more than enough. */
+#define MAX_CAPSULES 32
+
+/* 4 should be enough, but 8 won't hurt. */
+#define CAPSULE_ALIGNMENT 8
+
+/*
+ * A helper structure which bundles physical block address with its data.
+ *
+ * Usage:
+ *  1) Set .self to physical address
+ *  2) Check block's address with is_good_block()
+ *  3) Use load_block() to fetch or store_block() to update data
+ */
+struct block_descr {
+	uint64_t self;
+
```

Comment:
```suggestion
```

Reply:
Is it intentional?

Reply:
It is intentional to separate real fields from location.  Can add a comment to make it clearer.

Reply:
That would help, otherwise it simply looks as a redundant newline

Reply:
https://github.com/Dasharo/coreboot/compare/0b4ea6501c3e8d88ccecb8c880b259c4b8949e43..a43a5cf6af820f6803bf3c6f35371c0c2a4b1d89#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1b

Also moved savedefconfig commit to https://github.com/Dasharo/coreboot/pull/519 and changed its base to `dasharo`.

Reply:
Thank you.